lang/src/accounts/account.rs

```diff
@@ -332,12 +340,12 @@ where
     }
 }
 
-impl<'info, T: AccountSerialize + AccountDeserialize + Owner + Clone> AccountsExit<'info>
+impl<'info, T: AccountSerialize + AccountDeserialize + Clone> AccountsExit<'info>
     for Account<'info, T>
 {
     fn exit(&self, program_id: &Pubkey) -> Result<()> {
         // Only persist if the owner is the current program and the account is not closed.
-        if &T::owner() == program_id && !crate::common::is_closed(&self.info) {
+        if self.info.owner == program_id && !crate::common::is_closed(&self.info) {
```

Comment:
this is actually wrong, we are using a fixed value here for the owner of the account, `T::owner()` is not dynamically assigned. The check as to whether the current owner is the program_id is done in `is_closed`

Changing the logic for the checks may be ok but i'm just making note of the fact that these changes are not equivalent.

Reply:
Correct, this change is required for an interface account, however, since there is no one `owner`.

Is there a reason for checking this from the static value rather than what's actually on the account? This is checking to see if the changes to the account should be written, which is only when the given `program_id` owns *this* account.

Thankfully, it's not possible thanks to the `try_from` checks, but if you could cast the underlying account into another type with a different owner, you could trick `exit` into trying to serialize an account with the wrong owner, so I would argue that this is more correct.

Reply:
It's just sort of how accounts work. It's not possible load up an `Account` unless it's owned by the program or immediately initialized to be owned by the program. 

So when `exit`ing the state of the account ought to be 
- the current program _is_ the owner 
- the account has been safely closed 

Whereas with this change any other program owner is now valid. 

Before this change if account's owner is this program then it has to be serialized, with the only exception being that the account has just been closed.

I'm still just working out the logic myself but it I think boils down to being really strict about `Account` only ever being owned by this program or the system program. At least that's how it was before.

Reply:
> Whereas with this change any other program owner is now valid. So it's now possible to take an Account and change its owner to some random program.

Ah interesting, I see why it's done this way now, thanks! In general, you can safely change the owner of an account to any other program, if the data is zero'ed out. But as you say, non-program owner can't change their owner, so here's the problematic scenario:

* Program A owns account 1, program B owns account 2
* At the start of A, accounts 1 and 2 are deserialized, ownership checks performed
* Program A CPIs into program B with account 2
* Program B zeroes out account 2 and reassigns ownership to program A
* At the end of program B, nothing happens because ownership has changed
* At the end of program A, it writes changes to account 1, and rewrites the old data into account 2, creating a spoofed account, and at the very least bricking account 2, since it will never pass a deserialization check ever again, 

With interfaces, this can be very nasty, especially if an interface calls into the same interface. You can reassign in the nested interface program, and then the caller program re-initializes that account for itself, creating a real spoof.

My first thought is to separately store the owner for interface accounts so we can properly check the owner during `exit`, but maybe we can do better. What do you think?

Reply:
> Maybe it's ok that interface accounts can change owners inside of a function but it's up to you / what other expect the functionality to support. Like maybe people want a migration function or something.

No no, we don't want anything like that, trust me :sweat_smile: I've made a change to store the owner separately and check against that. I've also reverted the `Account` owner check back to use `T::owner()`

Reply:
Updated the check in [639775d](https://github.com/coral-xyz/anchor/pull/2386/commits/639775d61e046c51b7d208356a547a4ad6df6bc5)